src/privacy/mod_privacy_riak.erl

```diff
@@ -0,0 +1,173 @@
+    case mongoose_riak:fetch_type(?BKT_LISTS_NAMES(HostType), key(LUser, LServer)) of
+        {ok, Set} ->
+            riakc_set:value(Set);
+        {error, {notfound, set}} ->
+            [];
+        Err ->
+            ?LOG_ERROR(#{what => privacy_get_list_names_only_failed,
+                         user => LUser, server => LServer, reason => Err}),
+            []
+    end.
+
+get_privacy_list(HostType, LUser, LServer, Name) ->
+    case mongoose_riak:get(?BKT_LISTS(HostType), key(LUser, LServer, Name)) of
+        {ok, Obj} ->
+            Val = binary_to_term(riakc_obj:get_value(Obj)),
+            {ok, Val};
+        {error, notfound} ->
+            {error, not_found};
+        Err ->
+            ?LOG_ERROR(#{what => privacy_get_list_names_only_failed,
+                         user => LUser, server => LServer, list_name => Name,
+                         reason => Err}),
+            Err
+    end.
+
+forget_default_list(HostType, LUser, LServer) ->
+    mongoose_riak:delete(?BKT_DEFAULT_LIST(HostType), key(LUser, LServer)).
+
+set_default_list(HostType, LUser, LServer, Name) ->
+    case mongoose_riak:get(?BKT_LISTS(HostType), key(LUser, LServer, Name)) of
+        {ok, _} ->
+            % create or update
+            Obj = riakc_obj:new(?BKT_DEFAULT_LIST(HostType), key(LUser, LServer), Name),
+            mongoose_riak:put(Obj);
+        % in case list name is not found
+        {error, notfound} ->
+            {error, not_found};
+        Err -> Err
+    end.
+
+remove_privacy_list(HostType, LUser, LServer, Name) ->
+    mongoose_riak:delete(?BKT_LISTS(HostType), key(LUser, LServer, Name)),
+
+    case mongoose_riak:fetch_type(?BKT_LISTS_NAMES(HostType), key(LUser, LServer)) of
+        {ok, S1} ->
+            S2 = riakc_set:del_element(Name, S1),
+            mongoose_riak:update_type(?BKT_LISTS_NAMES(HostType), key(LUser, LServer), riakc_set:to_op(S2));
+        Err ->
+            ?LOG_ERROR(#{what => privacy_remove_privacy_list_failed,
+                         user => LUser, server => LServer, list_name => Name,
+                         reason => Err}),
+            Err
+    end.
+
+replace_privacy_list(HostType, LUser, LServer, Name, List) ->
+    % store privacy-list content
+    BinaryList = term_to_binary(List),
+    Obj = riakc_obj:new(?BKT_LISTS(HostType), key(LUser, LServer, Name), BinaryList),
+    mongoose_riak:put(Obj),
+
+    % add new list name to user privacy-lists set
+    S = riakc_set:add_element(Name, riakc_set:new()),
+    mongoose_riak:update_type(?BKT_LISTS_NAMES(HostType), key(LUser, LServer), riakc_set:to_op(S)).
+
+remove_user(HostType, LUser, LServer) ->
+    forget_default_list(HostType, LUser, LServer),
+    lists:foreach(
+        fun(Name) -> remove_privacy_list(HostType, LUser, LServer, Name) end,
+        get_list_names_only(HostType, LUser, LServer)
+    ),
+    mongoose_riak:delete(?BKT_LISTS_NAMES(HostType), key(LUser, LServer)).
+
+%% TODO
```

Comment:
This TODO stuff should be implemented in other PR?

Reply:
Yeah, unless anybody know how to implement this for riak, in riak keys are the whole jid as a binary, I don't know how to tell riak to remove keys by a pattern-match 🤔

Reply:
For this case when those keys will not deleted from riak - what will happens if we will try to put duplicate keys? It's can be provide tests for such a case and make sure that it does not break logic in other places?
Other question is - if we do not delete the keys, does this mean that the keys will be stored forever and gradually clutter up the space?

Reply:
Keys are the entire jid of the user, so reinserting it would mean I guess reinserting this user's configuration. This hasn't changed, if the old logic was correct, the new one is as well, all that has changed is the key name.
This new hook `remove_domain` is related to multitenancy, and no modules had it a month ago. I just don't know how to implement such thing for riak in this case, but I'm pretty confident riak and mod_privacy is not used by anyone, so I wasn't in a hurry to implement this new callback. Still, we should investigate how to delete keys based on a pattern for riak.